File: api/core/hunt/ingestion/runner.py
"""Parallel runner — fans a query out to many tools concurrently.

Uses a thread pool. Per-tool failures are caught and turned into
empty finding lists, so one slow/broken upstream cannot poison a
hunt. A global per-run timeout bounds the worst case.
"""
from __future__ import annotations

import concurrent.futures
import logging
from typing import Any, Iterable, List, Mapping, Sequence

from hunt.ingestion.base import Finding, ToolFunction

log = logging.getLogger(__name__)


class ParallelRunner:
    def __init__(self, tools: Sequence[ToolFunction], max_workers: int = 6):
        self._tools = list(tools)
        self._max_workers = max(1, max_workers)

    def run(self, query: Mapping[str, Any]) -> List[Finding]:
        findings: List[Finding] = []
        with concurrent.futures.ThreadPoolExecutor(
            max_workers=self._max_workers,
            thread_name_prefix="hunt-tool",
        ) as ex:
            futures = {
                ex.submit(_safe_run, tool, dict(query)): tool for tool in self._tools
            }
            for fut in concurrent.futures.as_completed(futures, timeout=120):
                tool = futures[fut]
                try:
                    res = fut.result(timeout=30)
                except concurrent.futures.TimeoutError:
                    log.warning("Tool %s timed out", tool.name)
                    continue
                except Exception as exc:  # noqa: BLE001
                    log.warning("Tool %s crashed: %s", tool.name, exc)
                    continue
                findings.extend(res)
        return findings


def _safe_run(tool: ToolFunction, query: Mapping[str, Any]) -> List[Finding]:
    try:
        return tool.run(query)
    except Exception as exc:  # noqa: BLE001
        log.warning("Tool %s failed: %s", tool.name, exc)
        return []


def run_investigation(tools: Sequence[ToolFunction], query: Mapping[str, Any]) -> List[Finding]:
    return ParallelRunner(tools).run(query)

File: lib/tools/runner.ts
/**
 * Parallel runner — fans a query out to many tools concurrently.
 *
 * Per-tool failures are caught and turned into empty finding lists
 * (or a single error finding, depending on which layer failed), so
 * one slow/broken upstream cannot poison a hunt.
 *
 * The concurrency cap defaults to 6 (matches the Python version).
 * Bounded by a per-tool timeout (ToolFunction.per_request_timeout_ms).
 */

import type { Finding, ToolFunction, ToolQuery } from "./base";

export interface RunResult {
  tool: string;
  findings: Finding[];
  duration_ms: number;
  error?: string;
}

export async function runTools(
  tools: ToolFunction[],
  query: ToolQuery,
  concurrency = 6
): Promise<RunResult[]> {
  const results: RunResult[] = [];
  let cursor = 0;
  const next = async (): Promise<void> => {
    const i = cursor++;
    if (i >= tools.length) return;
    const tool = tools[i];
    const t0 = Date.now();
    try {
      const findings = await tool.run(query);
      results.push({
        tool: tool.name,
        findings,
        duration_ms: Date.now() - t0,
      });
    } catch (e) {
      results.push({
        tool: tool.name,
        findings: [],
        duration_ms: Date.now() - t0,
        error: String(e).slice(0, 240),
      });
    }
    await next();
  };
  const workers = Array.from(
    { length: Math.min(concurrency, tools.length) },
    () => next()
  );
  await Promise.all(workers);
  return results;
}
